bundle/edu.gemini.sp.vcs/src/main/scala/edu/gemini/sp/vcs2/VcsServer.scala

```diff
@@ -179,16 +180,40 @@ class VcsServer(odb: IDBDatabaseService) { vs =>
     }
 
   private def accessControlled[A](id: SPProgramID, user: Set[Principal])(body: => VcsAction[A]): VcsAction[A] =
-    hasPermission(new ProgramPermission.Read(id), user) >>= { hp =>
+    hasPermission(ProgramPermission.Read(id), user) >>= { hp =>
       if (hp) body
       else {
         VcsServer.Log.info(s"VCS op forbidden: pid=$id, user=[${user.toList.mkString(", ")}]")
         VcsAction.fail(Forbidden(s"You don't have permission to access program '$id'"))
       }
     }
 
-  private def locked[A](k: SPNodeKey, lock: SPNodeKey => Unit, unlock: SPNodeKey => Unit)(body: => VcsAction[A]): VcsAction[A] =
-    VcsAction(lock(k)) >> { try { body} finally { unlock(k) } }
+  private def locked[A](k: SPNodeKey, lock: SPNodeKey => Unit, unlock: SPNodeKey => Unit)(body: => VcsAction[A]): VcsAction[A] = {
+    // Run the `body` while a lock is held.  If the releasing thread differs
+    // from the locking thread, at least warn about it before it blows up.
+    val a = new AtomicLong()
+
+    def recordAndLock(): Unit = {
+      a.set(Thread.currentThread.getId)
+      lock(k)
+    }
+
+    def checkAndUnlock(): Unit = {
+      val lockingThreadId   = a.get
+      val unlockingThreadId = Thread.currentThread.getId
+      if (lockingThreadId != unlockingThreadId) {
+        VcsServer.Log.severe(s"Unlocking thread ($unlockingThreadId) differs from locking thread ($lockingThreadId).  This will fail.")
+      }
+      unlock(k)
+    }
+
+    VcsAction(recordAndLock()) >>
+      // catch exceptions in the calculation of `body` itself
+      (try { body } catch { case e: Throwable => VcsFailure.vcsException(e).liftVcs[A] })
+        .run
+        .onFinish(_ => Task.delay(checkAndUnlock()))
+        .liftVcs
```

Comment:
This little block holds the money change, insuring that the lock is released after the action is actually executed.